go/ct/spc/specification.go

```diff
@@ -1475,6 +1475,123 @@ func getAllRules() []Rule {
 		},
 	})...)
 
+	// --- SELFDESTRUCT ---
+
+	rules = append(rules, rulesFor(instruction{
+		op:        SELFDESTRUCT,
+		name:      "_istanbul_cold",
+		staticGas: 5000,
+		pops:      1,
+		pushes:    0,
+		conditions: []Condition{
+			Eq(ReadOnly(), false),
+			IsRevision(R07_Istanbul),
+			HasNotSelfDestructed(ContractAccount()),
+			IsAddressCold(Param(0)),
+		},
+		parameters: []Parameter{
+			AddressParameter{},
+		},
+		effect: func(s *st.State) {
+			selfDestructEffect(s, vm.Gas(0), vm.Gas(25000), vm.Gas(24000))
+		},
+	})...)
+
+	rules = append(rules, rulesFor(instruction{
+		op:        SELFDESTRUCT,
+		name:      "_istanbul_warm",
+		staticGas: 5000,
+		pops:      1,
+		pushes:    0,
+		conditions: []Condition{
+			Eq(ReadOnly(), false),
+			IsRevision(R07_Istanbul),
+			HasNotSelfDestructed(ContractAccount()),
+			IsAddressWarm(Param(0)),
+		},
+		parameters: []Parameter{
+			AddressParameter{},
+		},
+		effect: func(s *st.State) {
+			selfDestructEffect(s, vm.Gas(0), vm.Gas(0), vm.Gas(24000))
+		},
+	})...)
+
+	rules = append(rules, rulesFor(instruction{
+		op:        SELFDESTRUCT,
+		name:      "_berlin_cold",
+		staticGas: 5000,
+		pops:      1,
+		pushes:    0,
+		conditions: []Condition{
+			Eq(ReadOnly(), false),
+			IsRevision(R09_Berlin),
+			HasNotSelfDestructed(ContractAccount()),
+			IsAddressCold(Param(0)),
+		},
+		parameters: []Parameter{
+			AddressParameter{},
+		},
+		effect: func(s *st.State) {
+			selfDestructEffectWithWarming(s, vm.Gas(2600), vm.Gas(0), vm.Gas(24000))
+		},
+	})...)
+
+	rules = append(rules, rulesFor(instruction{
+		op:        SELFDESTRUCT,
+		name:      "_berlin_warm",
+		staticGas: 5000,
+		pops:      1,
+		pushes:    0,
+		conditions: []Condition{
+			Eq(ReadOnly(), false),
+			IsRevision(R09_Berlin),
+			HasNotSelfDestructed(ContractAccount()),
+			IsAddressWarm(Param(0)),
+		},
+		parameters: []Parameter{
+			AddressParameter{},
+		},
+		effect: func(s *st.State) {
+			selfDestructEffectWithWarming(s, vm.Gas(0), vm.Gas(0), vm.Gas(24000))
+		},
+	})...)
+
+	rules = append(rules, rulesFor(instruction{
+		op:        SELFDESTRUCT,
+		name:      "_london",
+		staticGas: 5000,
+		pops:      1,
+		pushes:    0,
+		conditions: []Condition{
+			Eq(ReadOnly(), false),
+			IsRevision(R10_London),
+			HasNotSelfDestructed(ContractAccount()),
+			IsAddressWarm(Param(0)),
+		},
+		parameters: []Parameter{
+			AddressParameter{},
+		},
+		effect: func(s *st.State) {
+			selfDestructEffect(s, vm.Gas(0), vm.Gas(0), vm.Gas(0))
+		},
+	})...)
+
+	rules = append(rules, Rule{
+		Name: "selfdestruct_staticcall",
+		Condition: And(
+			Eq(ReadOnly(), true),
+			AnyKnownRevision(),
+			Eq(Status(), st.Running),
+			Eq(Op(Pc()), SELFDESTRUCT),
+			Ge(Gas(), 5000),
+		),
+		Parameter: []Parameter{
+			AddressParameter{},
+		},
+		Effect: FailEffect(),
+	})
```

Comment:
I would suspect that this leaves the the SELFDESTRUCT operation under-specified. For instance, a state that is read-only, running, with a SELFDESTRUCT operation and 4000 gas has no associated rule.

I am also surprised that this is not a case that is tested. The test-state generation should produce a state where all conditions as listed here are covered, except for the Gas.